index.js

```diff
@@ -34,19 +38,201 @@ class ServerlessPlugin {
+
+        let iamRoleStatements = this.serverless.service.provider.iamRoleStatements;
+        const roleDetails = {
+            'Effect': 'Allow',
+            'Action': 's3:GetObject',
+            'Resource': [
+                `${this.bucketName}/vendors/*`
+            ]
+        };
+
+        if(typeof iamRoleStatements !== 'undefined' && iamRoleStatements) {
+            if(iamRoleStatements.indexOf(roleDetails) === -1) {
+                iamRoleStatements[iamRoleStatements.length] = roleDetails;
+            }
+        } else {
+            this.serverless.service.provider.iamRoleStatements = [
+                roleDetails
+            ];
+        }
+
+        console.log(this.serverless.service.provider.environment);
+
+        console.log('Bref: Vendor separation done!');
+    }
+
+    async createZipFile() {
+        const filePath = 'vendor.zip';
+
+        return await new Promise((resolve, reject) => {
+            const archiver = require(process.mainModule.path + '/../node_modules/archiver');
+            const output = this.fs.createWriteStream(filePath);
+            const archive = archiver('zip', {
+                zlib: { level: 9 } // Sets the compression level.
+            });
+
+            console.log(`Bref: Creating ${filePath} archive...`);
+
+            archive.pipe(output);
+            archive.directory('vendor/', false);
+            archive.finalize();
+
+            output.on('close', () => {
+                console.log(`Bref: Created ${filePath} with ${archive.pointer()} total bytes.`);
+                resolve();
+            });
+
+            output.on('end', () => {
+                console.log('Bref: Archiver data stream has been drained');
+            });
+
+            archive.on('warning', err => {
+                if (err.code === 'ENOENT') {
+                    // log warning
+                    console.warn('Bref: Archiver warning', err);
+                } else {
+                    // throw error
+                    console.error('Bref: Archiver warning', err);
+                    reject(err);
+                }
+            });
+
+            archive.on('error', err => {
+                console.error('Bref: Archiver error', err);
+                reject(err);
+            });
+        })
+            .then(() => {
+                const crypto = require('crypto');
+
+                return new Promise(resolve => {
+                    const hash = crypto.createHash('md5');
+                    this.fs.createReadStream(filePath).on('data', data => hash.update(data)).on('end', () => resolve(hash.digest('hex')));
+                });
+            })
+            .then(hash => {
+                return hash;
+            })
+            .catch(err => {
+                throw new Error(`Failed to create zip file "${filePath}": ${err.message}`);
+            });
+    }
+
+    async uploadZipToS3(zipFile) {
+        const bucketObjects = await this.provider.request('S3', 'listObjectsV2', {
```

Comment:
Would it be simpler to use [`headObject`](https://docs.aws.amazon.com/AmazonS3/latest/API/API_HeadObject.html) to check if the object already exist?

Reply:
That's a really good idea. I'll implement this.